File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.js
const fs = require("fs");

let graph = [];
let virus = [];
let que = [];
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
// Assuming the file content is in 'input.txt'
fs.readFile(filePath, "utf8", function (err, data) {
  if (err) {
    console.error(`Error reading file: ${err}`);
    return;
  }

  // Split the file content by newline
  const lines = data.split("\n");
  let lineIndex = 0;

  // Process the first line to get n and k
  let [n, k] = lines[lineIndex++].split(" ").map(Number);
  graph = lines
    .slice(lineIndex, lineIndex + n)
    .map((line) => line.split(" ").map(Number));
  lineIndex += n;

  // Process the second line to get s, x, and y
  let [s, x, y] = lines[lineIndex++].split(" ").map(Number);

  // Process the virus data
  for (let r = 0; r < graph.length; r++) {
    for (let c = 0; c < graph[0].length; c++) {
      if (graph[r][c]) {
        virus.push([graph[r][c], r, c]);
      }
    }
  }
  virus.sort((a, b) => a[0] - b[0]);

  // Initialize the queue with virus data
  for (let v of virus) {
    que.push([v[1], v[2], 0]);
  }

  // The rest of the code remains the same
  const dr = [-1, 1, 0, 0];
  const dc = [0, 0, -1, 1];

  function bfs() {
    while (que.length) {
      let [r, c, time] = que.shift();

      if (time === s) {
        continue;
      }

      for (let idx = 0; idx < 4; idx++) {
        let nr = r + dr[idx],
          nc = c + dc[idx];
        if (nr < 0 || nr >= graph.length || nc < 0 || nc >= graph[0].length) {
          continue;
        }
        if (graph[nr][nc] === 0) {
          graph[nr][nc] = graph[r][c];
          que.push([nr, nc, time + 1]);
        }
      }
    }
  }

  bfs();
  console.log(graph[x - 1][y - 1]);
});

File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.py
## https://www.acmicpc.net/problem/18405

import sys
input = sys.stdin.readline
from collections import deque

graph = []
n,k =  map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
s,x,y = map(int, input().split())
que = deque()
virus = []
## print(graph, s,x,y)



    
for r in range(len(graph)):
    for c in range(len(graph[0])):
        if (graph[r][c]):
            virus.append([graph[r][c],r,c])
virus.sort()

for v,r,c in virus:
    que.append([r,c,0])

dr = [-1,1,0,0]
dc = [0,0,-1,1]


def bfs():
    while que:
        r,c, time = que.popleft()
        
        if time ==s:
            continue

        for idx in range(4):
            nr, nc = r + dr[idx], c + dc[idx]
            if not(0<=nr<n) or not(0<=nc<n): ## 범위 벗어나는 경우
                continue
            if graph[nr][nc]==0:
                graph[nr][nc]=graph[r][c]
                que.append([nr, nc, time+1])



bfs()            
print(graph[x-1][y-1])